mavros_extras/src/plugins/optical_flow.cpp

```diff
@@ -0,0 +1,209 @@
+ */
+class OpticalFlowPlugin : public plugin::Plugin
+{
+public:
+  explicit OpticalFlowPlugin(plugin::UASPtr uas_)
+  : Plugin(uas_, "optical_flow"),
+    ranger_fov(0.0),
+    ranger_min_range(0.3),
+    ranger_max_range(5.0)
+  {
+    enable_node_watch_parameters();
+
+    node_declare_and_watch_parameter(
+      "frame_id", "optical_flow", [&](const rclcpp::Parameter & p) {
+        frame_id = p.as_string();
+      });
+
+    /**
+     * @note Default rangefinder is Maxbotix HRLV-EZ4
+     * This is a narrow beam (60cm wide at 5 meters,
+     * but also at 1 meter). 6.8 degrees at 5 meters, 31 degrees
+     * at 1 meter
+     */
+    node_declare_and_watch_parameter(
+      "ranger_fov", 0.119428926, [&](const rclcpp::Parameter & p) {
+        ranger_fov = p.as_double();
+      });
+
+    node_declare_and_watch_parameter(
+      "ranger_min_range", 0.3, [&](const rclcpp::Parameter & p) {
+        ranger_min_range = p.as_double();
+      });
+
+    node_declare_and_watch_parameter(
+      "ranger_max_range", 5.0, [&](const rclcpp::Parameter & p) {
+        ranger_max_range = p.as_double();
+      });
+
+    flow_pub = node->create_publisher<mavros_msgs::msg::OpticalFlow>(
+      "~/raw/optical_flow", 10);
+    range_pub = node->create_publisher<sensor_msgs::msg::Range>("~/ground_distance", 10);
+
+    flow_sub = node->create_subscription<mavros_msgs::msg::OpticalFlow>(
+      "~/raw/send", 1, std::bind(
+        &OpticalFlowPlugin::send_cb, this,
+        _1));
+  }
+
+  Subscriptions get_subscriptions() override
+  {
+    return {
+      make_handler(&OpticalFlowPlugin::handle_optical_flow)
+    };
+  }
+
+private:
+  std::string frame_id;
+
+  double ranger_fov;
+  double ranger_min_range;
+  double ranger_max_range;
+
+  rclcpp::Publisher<mavros_msgs::msg::OpticalFlow>::SharedPtr flow_pub;
+  rclcpp::Publisher<sensor_msgs::msg::Range>::SharedPtr range_pub;
+  rclcpp::Subscription<mavros_msgs::msg::OpticalFlow>::SharedPtr flow_sub;
+
+  void handle_optical_flow(
+    const mavlink::mavlink_message_t * msg [[maybe_unused]],
+    mavlink::common::msg::OPTICAL_FLOW & flow,
+    plugin::filter::SystemAndOk filter [[maybe_unused]])
+  {
+    auto header = uas->synchronized_header(frame_id, flow.time_usec);
+
+    /**
+     * Raw message with axes mapped to ROS conventions and temp in degrees celsius.
+     *
+     * The optical flow camera is essentially an angular sensor, so conversion is like
+     * gyroscope. (aircraft -> baselink)
+     */
+    auto int_flow = ftf::transform_frame_aircraft_baselink(
+      Eigen::Vector3d(
+        flow.flow_x,
+        flow.flow_y,
+        0.0));
+    auto int_flow_comp_m_xy = ftf::transform_frame_aircraft_baselink(
+      Eigen::Vector3d(
+        flow.flow_comp_m_x,
+        flow.flow_comp_m_y,
+        0.0));
+    auto int_flow_rate_xy = ftf::transform_frame_aircraft_baselink(
+      Eigen::Vector3d(
+        flow.flow_rate_x,
+        flow.flow_rate_y,
+        0.0));
```

Comment:
After that conversion i'm not sure that we can just keep XY. Maybe better Vector3?

Reply:
The z component of the vector is not used in the optical flow algorithm. The function `ftf::transform_frame_aircraft_baselink`, which multiplies the vector by `quaternion_from_rpy(M_PI, 0.0, 0.0)`, it becomes clear that the z component does not contribute to the algorithm. Consequently, it is recommended to preserve only the XY components of the vector and explicitly set the z component to 0.0 during the conversion process. This approach aligns with the requirements of the algorithm and ensures proper functioning.